File: content/snippets/go/design_patterns/multiton.go
package main

import "fmt"

var instances = map[string]*Multiton{}

type Multiton struct{ Key string }

func Get(key string) *Multiton {
	if v, ok := instances[key]; ok {
		return v
	}
	m := &Multiton{Key: key}
	instances[key] = m
	return m
}
func main() {
	a := Get("a")
	b := Get("b")
	fmt.Println(a.Key, b.Key)
}

File: content/snippets/cpp/design_patterns/multiton.cpp
#include <map>
#include <memory>
#include <string>
#include <iostream>

class Multiton {
    static std::map<std::string, std::shared_ptr<Multiton>> instances;
public:
    std::string name;
    Multiton(std::string n):name(n){}
    static std::shared_ptr<Multiton> get(const std::string &k){
        auto it = instances.find(k);
        if (it!=instances.end()) return it->second;
        auto v = std::make_shared<Multiton>(k);
        instances[k]=v; return v;
    }
};
std::map<std::string, std::shared_ptr<Multiton>> Multiton::instances;
int main(){
    auto a = Multiton::get("a");
    auto b = Multiton::get("b");
    std::cout<<a->name<<" "<<b->name<<"\n";
}
